test/integration/http_integration.cc

```diff
@@ -251,7 +250,8 @@ Network::ClientConnectionPtr HttpIntegrationTest::makeClientConnectionWithOption
       std::make_shared<quic::QuicCryptoClientConfig>(
           std::make_unique<Quic::EnvoyQuicProofVerifier>(
               quic_transport_socket_factory_ref.sslCtx()),
-          std::make_unique<quic::QuicClientSessionCache>()),
+          dynamic_cast<Quic::PersistentQuicInfoImpl&>(*quic_connection_persistent_info_)
```

Comment:
oh hm, if we're having to do dynamic casts, what do we think of instead turning this into a proper Envoy interface, and having getters and setters.  I think if we #define away the includes and the methods for Non-http/3 builds it could work to pass around the real deal but I'm not 100% sure.

Reply:
As I understand it, we will never be able to get rid of ENVOY_ENABLE_QUIC because of the builds which do not include boringSSL (which QUICHE needs).

I wonder if we could do something a simple as :

```
#ifdef ENVOY_ENABLE_QUIC
#include ....
struct PersistentQuicInfo { ... };
#else
struct PersistentQuicInfo {};
#endif
```

In other words, make the struct completely empty when ENVOY_ENABLE_QUIC is not defined?

Reply:
> As I understand it, we will never be able to get rid of ENVOY_ENABLE_QUIC because of the builds which do not include boringSSL (which QUICHE needs).

If FIPS updates to have quic interfaces, we can build QUICHE on FIPS build even though it doesn't include boringSSL.

> I wonder if we could do something a simple as :
> 
> ```
> #ifdef ENVOY_ENABLE_QUIC
> #include ....
> struct PersistentQuicInfo { ... };
> #else
> struct PersistentQuicInfo {};
> #endif
> ```
> 
> In other words, make the struct completely empty when ENVOY_ENABLE_QUIC is not defined?

I would prefer to add a virtual interface over this because this way is even easier to mess up with build config. `QuicStatName` declared in similar way already caused a runtime crash before.

I tried adding those QUICHE interfaces to Http::QuicPersistentInfo and realized there would be a dependency circle between cluster_manager_interface and quic_platform. We probably need to merge/split more build targets to break this cycle:
```
 //source/common/config:utility_lib (c3499de1262d039dc135a623b7cf7d4c693662fb0cd4b1ad80ef6f6683b9b61c)
.-> //envoy/upstream:cluster_manager_interface (c3499de1262d039dc135a623b7cf7d4c693662fb0cd4b1ad80ef6f6683b9b61c)
|   @com_github_google_quiche//:quic_core_config_lib (c3499de1262d039dc135a623b7cf7d4c693662fb0cd4b1ad80ef6f6683b9b61c)
|   @com_github_google_quiche//:quic_core_crypto_encryption_lib (c3499de1262d039dc135a623b7cf7d4c693662fb0cd4b1ad80ef6f6683b9b61c)
|   @com_github_google_quiche//:quic_core_data_lib (c3499de1262d039dc135a623b7cf7d4c693662fb0cd4b1ad80ef6f6683b9b61c)
|   @com_github_google_quiche//:quic_core_packets_lib (c3499de1262d039dc135a623b7cf7d4c693662fb0cd4b1ad80ef6f6683b9b61c)
|   @com_github_google_quiche//:quic_platform (c3499de1262d039dc135a623b7cf7d4c693662fb0cd4b1ad80ef6f6683b9b61c)
|   //source/common/quic/platform:quic_platform_impl_lib (c3499de1262d039dc135a623b7cf7d4c693662fb0cd4b1ad80ef6f6683b9b61c)
|   //source/common/filesystem:directory_lib (c3499de1262d039dc135a623b7cf7d4c693662fb0cd4b1ad80ef6f6683b9b61c)
|   //source/common/filesystem:directory_iterator_impl_lib_posix (c3499de1262d039dc135a623b7cf7d4c693662fb0cd4b1ad80ef6f6683b9b61c)
|   //source/common/api:os_sys_calls_lib (c3499de1262d039dc135a623b7cf7d4c693662fb0cd4b1ad80ef6f6683b9b61c)
|   //source/common/network:address_lib (c3499de1262d039dc135a623b7cf7d4c693662fb0cd4b1ad80ef6f6683b9b61c)
|   //source/common/network:socket_interface_lib (c3499de1262d039dc135a623b7cf7d4c693662fb0cd4b1ad80ef6f6683b9b61c)
|   //envoy/server:bootstrap_extension_config_interface (c3499de1262d039dc135a623b7cf7d4c693662fb0cd4b1ad80ef6f6683b9b61c)
|   //envoy/server:factory_context_interface (c3499de1262d039dc135a623b7cf7d4c693662fb0cd4b1ad80ef6f6683b9b61c)
`-- //envoy/upstream:cluster_manager_interface (c3499de1262d039dc135a623b7cf7d4c693662fb0cd4b1ad80ef6f6683b9b61c)
```

I stopped there because the change itself is already quite big and definitely worth a stand-alone change. Given that we are already down casting the persistent quic object without this PR, I added a TODO for future clean up.